squad/core/migrations/0093_populatehistoricalemailtemplate.py

```diff
@@ -0,0 +1,30 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.16 on 2018-10-24 19:49
+from __future__ import unicode_literals
+
+from django.db import migrations
+import simple_history.utils
+
+
+from squad.core.models import EmailTemplate
```

Comment:
usually it's not a good idea to use model classes directly in data migrations. instead, inside the function you do something like this:

```python
EmailTemplate = apps.get_model('core', 'EmailTemplate')
```

Reply:
hmm, it's not only a bad idea, in my experience it simply doesn't work. Creating object instances in migrations for some reason is broken. I guess this is intentional and the only proper way of creating object instances is what Antonio presented.

Reply:
Humm, I agree. But one thing this that's bugging me is that `EmailTemplate = apps.get_model('core', 'EmailTemplate')` retrieves a `__fake__.EmailTemplate` class, which doesn't contain `history` attribute (only 'id', 'name', 'subject' etc). It's just a fake object with all fields cloned into it. Thus the migration fails due to missing attribute 'history'.

Strangely, if using `from django.apps import apps as django_apps` and `EmailTemplate = apps.get_model('core', 'EmailTemplate')` then it'll return `squad.core.models.EmailTemplate`, which is the correct one.

In the [documentation](https://docs.djangoproject.com/en/1.7/topics/migrations/#data-migrations) the `apps` passed on `migrations.RunPython` should be the same as `django.apps.apps`, the app registry, but it's behaving differently. 

Any ideas on a workaround?

Reply:
> Humm, I agree. But one thing this that's bugging me is that `EmailTemplate = apps.get_model('core', 'EmailTemplate')` retrieves a `__fake__.EmailTemplate` class, which doesn't contain `history` attribute (only 'id', 'name', 'subject' etc). It's just a fake object with all fields cloned into it. Thus the migration fails due to missing attribute 'history'.
> 
> Strangely, if using `from django.apps import apps as django_apps` and `EmailTemplate = apps.get_model('core', 'EmailTemplate')` then it'll return `squad.core.models.EmailTemplate`, which is the correct one.

when running from tje django shell, this is the expected behavior: to return the "current" version of the model. Now, in the migration, it needs to return the model as it existed at the point where the migration was created, and that's why it's "fake" one: it is one that has the attributes and  relationships that the model had at that point.

not sure how to workaroudn this. a quick skim of the django-simple-history documentation doesn't seem to expect that bulk creation method being run from migrations. sorry, no ideas :-/

Reply:
Thanks for the input. I've came up with a workaround. It's not perfect, but I guess it solves our problem here. I looked it over simple-history's code and saw the internals of `bulk_history_create` and managed to replicate it on this migration. It basically creates one HistoricalEmailTemplate for each EmailTemplate. How does it sound to you?